lightclient/src/platform/wasm_socket.rs

```diff
@@ -0,0 +1,233 @@
+            data: VecDeque::with_capacity(16384),
+            waker: None,
+            callbacks: None,
+        }));
+
+        let open_callback = Closure::<dyn FnMut()>::new({
+            let inner = inner.clone();
+            move || {
+                let mut inner = inner.lock().expect("Mutex is poised; qed");
+                inner.state = ConnectionState::Opened;
+
+                if let Some(waker) = inner.waker.take() {
+                    waker.wake();
+                }
+            }
+        });
+        socket.set_onopen(Some(open_callback.as_ref().unchecked_ref()));
+
+        let message_callback = Closure::<dyn FnMut(_)>::new({
+            let inner = inner.clone();
+            move |event: web_sys::MessageEvent| {
+                let Ok(buffer) = event.data().dyn_into::<js_sys::ArrayBuffer>() else {
+                    panic!("Unexpected data format {:?}", event.data());
+                };
+
+                let mut inner = inner.lock().expect("Mutex is poised; qed");
+                let bytes = js_sys::Uint8Array::new(&buffer).to_vec();
+                inner.data.extend(bytes.into_iter());
+
+                if let Some(waker) = inner.waker.take() {
+                    waker.wake();
+                }
+            }
+        });
+        socket.set_onmessage(Some(message_callback.as_ref().unchecked_ref()));
+
+        let error_callback = Closure::<dyn FnMut(_)>::new({
+            let inner = inner.clone();
+            move |_| {
+                // Callback does not provide useful information, signal it back to the stream.
+                let mut inner = inner.lock().expect("Mutex is poised; qed");
+                inner.state = ConnectionState::Error;
+
+                if let Some(waker) = inner.waker.take() {
+                    waker.wake();
+                }
+            }
+        });
+        socket.set_onerror(Some(error_callback.as_ref().unchecked_ref()));
+
+        let close_callback = Closure::<dyn FnMut(_)>::new({
+            let inner = inner.clone();
+            move |_| {
+                let mut inner = inner.lock().expect("Mutex is poised; qed");
+                inner.state = ConnectionState::Closed;
+
+                if let Some(waker) = inner.waker.take() {
+                    waker.wake();
+                }
+            }
+        });
+        socket.set_onclose(Some(close_callback.as_ref().unchecked_ref()));
+
+        let callbacks = SendWrapper::new((
+            open_callback,
+            message_callback,
+            error_callback,
+            close_callback,
+        ));
+        inner.lock().expect("Mutex poised; qed").callbacks = Some(callbacks);
+
+        Ok(Self { inner })
+    }
+}
+
+impl AsyncRead for WasmSocket {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let mut inner = self.inner.lock().expect("Mutex is poised; qed");
```

Comment:
Another `Pending`/`Waker` thing:

Because this `WasmSocket` is behind an `Arc<Mutex<..>>`, a question I have is; what happens if we make, say, 3 copies of this, and then for each copy, in some different task, we `.read().await`?

Each copy would call `poll_read`, and then the waker given to each `poll_read` would be saved into `inner` to be woken up. But the wakers would overwrite eachother; 2 of the wakers would therefore never be used, and 2 of the three tasks would never make progress.

I had to consider this in some personal thing where I implemented a small MPSC style channel (https://github.com/jsdw/asyncified/blob/main/src/channel.rs), and there I had each clone of the sender allocate itself a new ID, and then I stored all wakers in a `HashMap<ID, waker>`) so that I would remove them at needed on cleanup, and wake them all up on progress.

That all said, `pub struct WasmSocket` is not itself clonable, so, does it need the `Arc` inside it at all? (or the mutex for that matter)? Because if you could clone the socket and read from it, different clones would be getting different parts of the output anyway which probably isn't what we want.

So maybe we can just remove the `Arc<Mutex>`? Or at least the `Arc`? And then the type system enforces that it can only be used in one place, with one waker, at a time.

Reply:
Ah, so the reason for the `Arc<Mutex<..>>` was because this thing needed sharing across some different closures below; I missed that bit! And when compiled to WASM the mutex will disappear anyway since it's single threaded.

I think given that the outer struct is not cloneable, I'd remove the `SendWrapper` thing (and crate) and just put the `Arc<Mutex<..>>` back (it feels cleaner than this `SendWrapper<Rc<RefCell<..>>>` thing). Just maybe a dev comment on `WasmSocket` that the thing cannot impl `Clone` would do I think :)